porespy/networks/__funcs__.py

```diff
@@ -77,85 +77,110 @@ def add_boundary_regions(regions=None, faces=['front', 'back', 'left',
-            if 'right' not in faces:
-                regions = regions[:-3, :]
-            if 'front' not in faces and 'bottom' not in faces:
-                regions = regions[:, 3:]  # y
-            if 'back' not in faces and 'top' not in faces:
-                regions = regions[:, :-3]
-        else:
-            print('add_boundary_regions works only on 2D and 3D images')
-        # ---------------------------------------------------------------------
-        # Make labels contiguous
-        regions = make_contiguous(regions)
-    else:
-        regions = regions
+    if faces is None:
+        return regions
+
+    if regions.ndim not in [2, 3]:
+        raise Exception("add_boundary_regions works only on 2D and 3D images")
+
+    if regions.ndim == 2:
+        if set(["bottom", "top"]).intersection(faces):
+            raise Exception("For 2D images, use 'left', 'right', 'front', 'back' labels")
+
+    # Map which image slice corresponds to which face
+    face_to_slice = {
+        "left": 0, "right": -1, "front": 0, "back": -1, "bottom": 0, "top": -1
+    }
+    # Map face label to corresponding axis
+    face_to_axis = {
+        "left": 0, "right": 0, "front": 1, "back": 1, "bottom": 2, "top": 2
+    }
+
+    ndim = regions.ndim
+    indices = []
+    # Note: pad_width format: each elem is [pad_before, pad_after]
+    pad_width = [[0, 0] for i in range(ndim)]
+    # 1. Create indices for boundary faces, ex. bottom" => [:, :, -1]
+    # Note: slice(None) is equivalent to ":" in fancy indexing, ex. [0, 0, :]
+    # 2. populate pad_width based on labels
+    for face in faces:
+        temp = [slice(None) for i in range(ndim)]
+        axis = face_to_axis[face]
+        plane = face_to_slice[face]
+        temp[axis] = plane
+        indices.append(tuple(temp))
+        pad_width[axis][plane] = 1      # Pad each face by 1 pixel
+
+    regions = np.pad(regions, pad_width=pad_width, mode="edge")
+
+    # Increment boundary regions to distinguish from internal regions
+    for idx in indices:
+        # Only increment non-background regions (i.e. != 0)
+        non_background = regions[idx] != 0
+        regions[idx][non_background] += regions.max()
+
+    # Remove connections between boundary regions
+    for idx in indices:
+        regions[idx] *= ~find_boundaries(regions[idx], mode="outer")
+
+    # Pad twice to make boundary regions 3-pixel thick -> required for marching_cube
+    regions = np.pad(regions, pad_width=pad_width, mode="edge")
+    regions = np.pad(regions, pad_width=pad_width, mode="edge")
+
+    # Convert pad-induced corners to 0
+    _zero_corners(regions, pad_width=3, faces=faces)
+
+    # Make labels contiguous
+    regions = make_contiguous(regions)
 
     return regions
 
 
+def _zero_corners(im, pad_width: int, faces: list):
+    r"""
+    Fills corners of a padded-image with 0, given a pad width and list of
+    faces such as ["left", "front"] at which the image has been padded.
+
+    Example
```

Comment:
You should make this example actual code by putting >>> for each line...then pytest will test it.

Reply:
Done